[WIP] Denmark marriage tool update - Content change request

pmanrubia opened the pull request.

sistephens commented:
Waiting to hear back from FCO on this one.

chrisroos commented:
@pmanrubia: Presumably you still don't want this reviewed until we've heard back from the department?

pmanrubia commented:
@chrisroos yes, thank you @chrisroos for the follow-up. I would still need to supersede this PR and write proper commits.

pmanrubia commented:
**Commits still need some polish and love. Not ready for review yet.**

This is the first iteration of changes for Denmark.

The flow for Denmark has three main paths: `living in UK`, `living in Denmark`, `living in third country`. This PR creates a new outcome for `living in UK`, `living in Denmark`, because the `third country` has not changed (it defaults to `:outcome_opposite_sex_marriage_in_consular_cni_countries_when_residing_in_third_country`)

I have tackled all the requested changes and the output for Denmark is what we would expect.

But, I a not happy with how the flow looks like. See code below for how we are covering the two first paths:

```ruby
elsif calculator.ceremony_country == 'denmark' && (calculator.resident_of_uk? || calculator.resident_of_ceremony_country?)
  outcome :outcome_ceremonies_in_denmark
```

To overcome this situation I am exploring these two options:

1. Create a new outcome for Denmark (third country), and rename the old one to something related to (resident_of_uk and resident_of_ceremony_country)
1. Update the flow and be explicit about Denmark and the third country

```ruby
elsif calculator.ceremony_country == 'denmark' && (calculator.resident_of_uk? || calculator.resident_of_ceremony_country?)
  outcome :outcome_ceremonies_in_denmark
elsif calculator.ceremony_country == 'denmark' && calculator.third_country?
outcome :outcome_opposite_sex_marriage_in_consular_cni_countries_when_residing_in_third_country
```
But None of them feels totally right to me. Any other suggestion, idea?

@chrisroos @ikennaokpala @floehopper  any idea, suggestion?

chrisroos commented:
I think the code snippet you posted looks OK, @pmanrubia:

```ruby
elsif calculator.ceremony_country == 'denmark' && (calculator.resident_of_uk? || calculator.resident_of_ceremony_country?)
  outcome :outcome_ceremonies_in_denmark
```

The only thing I'd change in there is the name of the outcome. I'd name it `outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark` to make it clear that we're only expecting it to be used in those circumstances.

You might notice that some of the other outcomes are similarly named but don't check for residency of UK or ceremony country (e.g. [poland](https://github.com/alphagov/smart-answers/blob/master/lib/smart_answer_flows/marriage-abroad.rb#L172) and [slovenia](https://github.com/alphagov/smart-answers/blob/master/lib/smart_answer_flows/marriage-abroad.rb#L174)). These outcomes are only reached when the user is resident of the UK or the ceremony country because of the ordering of the logic in the giant `if` statement (i.e. residents of a third country are sent to another outcome earlier on). I'd prefer that we were explicit with the logic required to reach an outcome to make it possible to reorder items in the `if` statement without worrying about affecting the behaviour.

Does that make sense? I'm afraid I'm not sure I've done a very good job of explaining what I mean so do let me know if what I've said is unclear.

pmanrubia commented:
@chrisroos Thank you for your feedback. 
I have renamed the outcome as per your suggestion (it reads better now)

The PR is ready. I would appreciate your review once you have some spare time.

chrisroos commented:
I'll take a look now, @pmanrubia.

pmanrubia commented:
Thank you @chrisroos.
I have just realised that 01a54e930a91e23bf29894c32c2810f12c035ff6 should go before cd38577b8c7781a37fd08a073190e7028e94cb96.

I must have messed up something with my last rebase. I am not going to force-push it now as you might have already done work on it.

Sorry about it.

chrisroos commented:
Hey @pmanrubia. I've reviewed this branch and am about to add a number of comments.

There's only one which is about a problem that I believe needs to be fixed (it should hopefully be clear from the comment which one it is).

I'll leave it to you to decide whether you think it's worth taking any action based on the rest of the comments.

Please shout if anything I've said isn't clear.

chrisroos commented:
__Including regression test artefact changes in the commits that cause them__

Where possible, I think we've been combining the changes to the regression test artefacts with the commit that causes the change. I think of this is analogous to changing both the test and behaviour in a single commit when modifying more traditional code.

In this branch, I think that would mean combining:

Commits 5 and 6:

* "Update Denmark outcome: Living in UK, opposite sex" (5dc594ae38692d7e57cb99bb42fc2c83afa30fa4)
* "Regenerate artefacts and checksums for marriage-abroad" (49765dcada1a396a7928700cfb5272c1a7a53f28)

Commits 8 and 9:

* "Update Denmark outcome: Living in UK, same sex" (a1c2f99c63b8df3d39699554ed17ec54192e5d2d)
* "Regenerate artefacts for marriage-abroad" (d3b7bccaed77e39543832ddbb709e6fddb4f57ed)

Commits 10 and 11:

* "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9b34a32c2056d5114475cf0253aac3dfe3)
* "Regenerate artefacts for marriage-abroad" (58792b252b49deec00e96d1b66e73557f3b1d26c)

Commits 12 and 13:

* "Update flow to use Denmark outcome: living Denmark, same sex" (cd38577b8c7781a37fd08a073190e7028e94cb96)
* "Regenerate artefacts for marriage-abroad" (01a54e930a91e23bf29894c32c2810f12c035ff6)

Commits 14 and 15:

* "Update fees for marriage-abroad in Denmark same sex" (1f68e65bf70c7f01821dafe8ab08a945105a6571)
* "Regenerate artefacts for marriage-abroad" (d0ba45fccd806df56794f9c1de914ff1bc399853)

Having written this, I've realised that it doesn't actually match our [regression test documentation](https://github.com/alphagov/smart-answers/blob/master/doc/regression-tests.md#how-to-get-the-regression-tests-to-pass-after-a-code-change). Does what I've said in this comment make sense, @floehopper? Or have we agreed to follow the documentation?

chrisroos commented:
__Updating marriage-abroad-responses-and-expected-results.yml__

A number of commits change the flow such that a set of responses lead to a different outcome. In general, I think we should be updating the marriage-abroad-responses-and-expected-results.yml in these commits. As with my comment about the regression test artefacts, I think this is analogous to the idea of updating tests and production code in a single commit. It has the advantage of allowing us to sanity check that the changes we've made have only effected the things we're expecting them to.

In this branch, the following commits are missing the related updates to marriage-abroad-responses-and-expected-results.yml:

* Commit 2. "Update flow to use Denmark outcome: living UK, opposite sex" (cd08fe7ae2d0ab718914c541fbd333c5905daf8a)
* Commit 7: "Update flow to use Denmark outcome: living UK, same sex" (756a1e6e8cd8d867c7540a2dfcc0d1e8d1b0cce7)
* Commit 10: "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9b34a32c2056d5114475cf0253aac3dfe3)
* Commit 12: "Update flow to use Denmark outcome: living Denmark, same sex" (cd38577b8c7781a37fd08a073190e7028e94cb96)

chrisroos commented:
__Combining commits 1 and 2__

I would've been inclined to combine the following commits, so that we're making use of the new template at the point it's introduced:

* "Duplicate outcome to extract new Outcome for Denmark" (146d44a49b7677df8e960fe50b024a5580146dcc)
* "Update flow to use Denmark outcome: living UK, opposite sex" (cd08fe7ae2d0ab718914c541fbd333c5905daf8a)

chrisroos commented:
__Commit 3. "Remove Denmark from outcome when Living in UK" (87b589bad200462dce673974c49b2d6c59736f89)__

It'd be good if the commit note explained that the change in the previous commit means that you'll never reach this outcome under these circumstances.

chrisroos commented:
__Commit 4. "Remove everything but Denmark from new outcome" (d172d2e787e94cfc4f270af02aee0aecb5996787)__

It's not immediately obvious (i.e. not without digging into the code) why some of these lines have been removed. It'd be good to explain their removal (i.e. because they return `false` for Denmark):

* `<% elsif calculator.opposite_sex_21_days_residency_required? %>`
* `<% if calculator.cni_posted_after_14_days? %>`
* `notary_public_ceremony_country?`

It also looks like there is more code that can be removed from this Denmark-specific outcome, e.g.:

* `<% if calculator.ceremony_country == 'denmark' %>`
* `<% if %w(kazakhstan macedonia russia).exclude?(calculator.ceremony_country) %>`
* `&& %w(finland).exclude?(calculator.ceremony_country)`?
* `calculator.country_without_consular_facilities? || calculator.ceremony_country == 'cote-d-ivoire'`

Note that neither of the above lists are exhaustive, they're just examples of things I spotted when I scanned the diff.

chrisroos commented:
__Commit 6: "Regenerate artefacts and checksums for marriage-abroad" (49765dcada1a396a7928700cfb5272c1a7a53f28)__

It looks like this contains some unrelated changes for Rwanda in marriage-abroad-responses-and-expected-results.yml.

chrisroos commented:
__Commit 7: "Update flow to use Denmark outcome: living UK, same sex" (756a1e6e8cd8d867c7540a2dfcc0d1e8d1b0cce7)__

It'd be good to update the name in this test too. From "go to cp or equivalent outcome" to "go to outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".

chrisroos commented:
__Commit 8: "Update Denmark outcome: Living in UK, same sex" (a1c2f99c63b8df3d39699554ed17ec54192e5d2d)__

__This one needs to be fixed!__

The addition of `<% if calculator.resident_of_ceremony_country? %>` appears to be unnecessary. It already exists in the template, and results in duplicate text in the regression test artefacts. You can see duplication in commits (search for "If you or your partner have been divorced"):

* Commit 11: "Regenerate artefacts for marriage-abroad" (58792b252b49deec00e96d1b66e73557f3b1d26c)
* Commit 13: "Regenerate artefacts for marriage-abroad" (01a54e930a91e23bf29894c32c2810f12c035ff6)

chrisroos commented:
__Commit 9: "Regenerate artefacts for marriage-abroad" (d3b7bccaed77e39543832ddbb709e6fddb4f57ed)__

It'd be great if you could explain in the commit note why you're happy to accept these changes to the regression test artefacts. Even if that's to say that you've copied them from the Google Doc.

The marriage-abroad-files.yml contains the markers of a merge conflict in this commit.

chrisroos commented:
__Commit 10: "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9b34a32c2056d5114475cf0253aac3dfe3)__

It'd be good to update the name of this test too. From "go to consular cni os outcome" to "go to outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".

chrisroos commented:
__Commit 13: "Regenerate artefacts for marriage-abroad" (01a54e930a91e23bf29894c32c2810f12c035ff6)__

The marriage-abroad-responses-and-expected-results.yml contains "outcome_ceremonies_in_denmark" yet the outcome is named "outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".

chrisroos commented:
__Commit 16: "Removes Denmark code from outcome." (3284e411e20eade3157d8a621f6d89525578b261)__

It'd be good to explain why this is OK in the commit note, i.e. because ceremonies in Denmark never reach that outcome.

chrisroos commented:
And that's everything for now, @pmanrubia. I realise there are a lot of comments but I hope they're more on the side of being useful than overwhelming!

pmanrubia commented:
@chrisroos: Thank you for your valuable review. 

I don’t usually force push on PRs to prevent losing comments, so next time would you mind providing feedback at the commit level when possible? It would be much easier for me to address changes and have a discussion / clarifications of your comments.

pmanrubia commented:
>In this branch, I think that would mean combining:

> Commits 5 and 6:
> ”Update Denmark outcome: Living in UK, opposite sex" (5dc594a)
> ”Regenerate artefacts and checksums for marriage-abroad" (49765dc)

> Commits 8 and 9:
> ”Update Denmark outcome: Living in UK, same sex" (a1c2f99)
> ”Regenerate artefacts for marriage-abroad" (d3b7bcc)

> Commits 10 and 11:
> ”Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9)
> ”Regenerate artefacts for marriage-abroad" (58792b2)

> Commits 12 and 13:
> ”Update flow to use Denmark outcome: living Denmark, same sex" (cd38577)
> ”Regenerate artefacts for marriage-abroad" (01a54e9)

> Commits 14 and 15:
> ”Update fees for marriage-abroad in Denmark same sex" (1f68e65)
> ”Regenerate artefacts for marriage-abroad" (d0ba45f)

I am not in favor of combining / merging them; my feeling is that artefacts have a different nature to tests, although I understand this might just a personal preference.
But I am not strongly attached to splitting them up, so if we all agree that we are going to merge them, then I am happy to do it. 
I guess we just need to make a decision and be consistent with it along the project, because as far as I know, as a team, we are not following the same approach when structuring commits.

cc/ @floehopper @leenagupte @ikennaokpala

pmanrubia commented:
> Updating marriage-abroad-responses-and-expected-results.yml

> A number of commits change the flow such that a set of responses lead to a different outcome. In general, I think we should be updating the marriage-abroad-responses-and-expected-results.yml in these commits. As with my comment about the regression test artefacts, I think this is analogous to the idea of updating tests and production code in a single commit. It has the advantage of allowing us to sanity check that the changes we've made have only effected the things we're expecting them to.

> In this branch, the following commits are missing the related updates to marriage-abroad-responses-and-expected-results.yml:

> Commit 2. "Update flow to use Denmark outcome: living UK, opposite sex" (cd08fe7)
> Commit 7: "Update flow to use Denmark outcome: living UK, same sex" (756a1e6)
> Commit 10: "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9)
> Commit 12: "Update flow to use Denmark outcome: living Denmark, same sex" (cd38577)

This is similar to my [comment above](https://github.com/alphagov/smart-answers/pull/2520#issuecomment-224648430). Those files were updated and merged later on:

Commit 2. "Update flow to use Denmark outcome: living UK, opposite sex" (cd08fe7)
It was updated in https://github.com/alphagov/smart-answers/pull/2520/commits/49765dcada1a396a7928700cfb5272c1a7a53f28

Commit 7: "Update flow to use Denmark outcome: living UK, same sex" (756a1e6)
It was updated in https://github.com/alphagov/smart-answers/pull/2520/commits/49765dcada1a396a7928700cfb5272c1a7a53f28

Commit 10: "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9)
It was updated in https://github.com/alphagov/smart-answers/pull/2520/commits/58792b252b49deec00e96d1b66e73557f3b1d26c

Commit 12: "Update flow to use Denmark outcome: living Denmark, same sex" (cd38577)
https://github.com/alphagov/smart-answers/pull/2520/commits/d3b7bccaed77e39543832ddbb709e6fddb4f57ed

My feeling is that those files are related to the regression tests, and if we prepare a commit small enough to be reviewed then it would be good to go. But again, if we all agree to update this file once we change the flow then let’s follow this convention for the incoming PRs.

pmanrubia commented:
> Commit 3. "Remove Denmark from outcome when Living in UK" (87b589b)

> It'd be good if the commit note explained that the change in the previous commit means that you'll never reach this outcome under these circumstances.

I have updated the commit notes in 79d8baeaa06f5efc5431c928959bbb66345eaca5

pmanrubia commented:
> It's not immediately obvious (i.e. not without digging into the code) why some of these lines have been removed. It'd be good to explain their removal (i.e. because they return false for Denmark):

```ruby
<% elsif calculator.opposite_sex_21_days_residency_required? %>
<% if calculator.cni_posted_after_14_days? %>
notary_public_ceremony_country?
```

> It also looks like there is more code that can be removed from this Denmark-specific outcome, e.g.:

> Note that neither of the above lists are exhaustive, they're just examples of things I spotted when I scanned the diff.

Thank you for detecting all the things I missed :+1:. I have also removed:

```ruby
%w(croatia italy russia).exclude?(calculator.ceremony_country) 
```

I have addressed these changes in https://github.com/alphagov/smart-answers/commit/bf9f9f097bcf4b5fa4d7770f7d59ea163ad6d9e1. I will fixup it up once we are happy with it.

I still need to update the commit notes as per your suggestion.

pmanrubia commented:
> Commit 6: "Regenerate artefacts and checksums for marriage-abroad" (49765dc)

> It looks like this contains some unrelated changes for Rwanda in marriage-abroad-responses-and-expected-results.yml.

It feels that after this commit https://github.com/alphagov/smart-answers/pull/2516/commits/b93c7a498d2b93f29c301843b0ffa2fca24cee7b in PR https://github.com/alphagov/smart-answers/pull/2516 we forgot to regenerate `marriage-abroad-responses-and-expected-results.yml`. We are currently addressing it: thank you @ikennaokpala for your work no this.

pmanrubia commented:
> Commit 7: "Update flow to use Denmark outcome: living UK, same sex" (756a1e6)

> It'd be good to update the name in this test too. From "go to cp or equivalent outcome" to "go to outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".


I have updated existing test from `go to cp or equivalent outcome` to `outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark` in 25e452f6477ce126dff42283319bd05d5a1d0502

pmanrubia commented:
> Commit 8: "Update Denmark outcome: Living in UK, same sex" (a1c2f99)

> This one needs to be fixed!

> The addition of <% if calculator.resident_of_ceremony_country? %> appears to be unnecessary. It already exists in the template, and results in duplicate text in the regression test artefacts. You can see duplication in commits (search for "If you or your partner have been divorced"):

> Commit 11: "Regenerate artefacts for marriage-abroad" (58792b2)
> Commit 13: "Regenerate artefacts for marriage-abroad" (01a54e9)

Thank you for spotting this duplication. 
I have addressed your comment in 783bdd11e9683b35cc166618a99c53be348555fe

pmanrubia commented:
> Commit 9: "Regenerate artefacts for marriage-abroad" (d3b7bcc)

> It'd be great if you could explain in the commit note why you're happy to accept these changes to the regression test artefacts. Even if that's to say that you've copied them from the Google Doc.

I might be missing something, but for `regression-tests` commits with not many changes, I can't see the benefit of making explicit something that is implicit in our work. I guess you might suggest it to prevent an automatic regeneration of artefacts without paying attention to its content, but still not sure if adding that text would help (after a few weeks the text could also be added without further thought).

In any case, I have updated the commits note with your suggestion in 660f549596d6ddec319cf27be94d06d7c473ffec

> The marriage-abroad-files.yml contains the markers of a merge conflict in this commit.

Thank you. Updated in 660f549596d6ddec319cf27be94d06d7c473ffec

pmanrubia commented:
> Commit 10: "Update flow to use Denmark outcome: living in Denmark, oppos sex" (0c6b3d9)

> It'd be good to update the name of this test too. From "go to consular cni os outcome" to "go to outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".

I have update the existing test from `go to consular cni os outcome` to `outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark` in 26f65702656bd5482ad5daee8c3135e84f0c4082

pmanrubia commented:
> Commit 13: "Regenerate artefacts for marriage-abroad" (01a54e9)

> The marriage-abroad-responses-and-expected-results.yml contains "outcome_ceremonies_in_denmark" yet the outcome is named "outcome_ceremonies_in_denmark_when_residing_in_uk_or_denmark".

Addressed in ff6bae73c7771ed9224a35e9d1cd45ef1703eeae

pmanrubia commented:
> Commit 16: "Removes Denmark code from outcome." (3284e41)

> It'd be good to explain why this is OK in the commit note, i.e. because ceremonies in Denmark never reach that outcome.

I have added `because ceremonies in Denmark never reach that outcome` to the commit notes in 2bb21955646c55cac29a2d78bc953b9a4ddda9f6

pmanrubia commented:
Sorry, I forgot this comment: 

> Combining commits 1 and 2

> I would've been inclined to combine the following commits so that we're making use of the new template at the point it's introduced:

> "Duplicate outcome to extract new Outcome for Denmark" (146d44a)
> "Update flow to use Denmark outcome: living UK, opposite sex" (cd08fe7)

Not sure about combining them: I would prefer being explicit  in the first commit where we explain that the outcome is copied as a previous step to migrate to the new one.

pmanrubia commented:
@chrisroos Thank you again for your feedback. It is much appreciated. I am going to close this PR and start a new one so we can continue the review.

chrisroos commented:
> I don’t usually force push on PRs to prevent losing comments, so next time would you mind providing feedback at the commit level when possible? It would be much easier for me to address changes and have a discussion / clarifications of your comments.

Hey @pmanrubia. I'm sorry that the way I added comments made it harder to review. How would you prefer I added commit level comments?

pmanrubia commented:
>> I don’t usually force push on PRs to prevent losing comments, so next time would you mind providing feedback at the commit level when possible? It would be much easier for me to address changes and have a discussion / clarifications of your comments.

> Hey @pmanrubia. I'm sorry that the way I added comments made it harder to review. How would you prefer I added commit level comments?

Thank you @chrisroos. I know we have lost in the past valuable commits because of force-pushing on a PR, and I know there are also comments that apply to the PR as a whole; but for comments related to a specific bit of code/commit, it would be handy to add the comment at the diff level?.  Something similar to this PR https://github.com/alphagov/smart-answers/pull/2502#discussion-diff-61885652. It's easy to add a `Line note` without having to manually include the context.
Does it make sense for you?

chrisroos commented:
> I might be missing something, but for regression-tests commits with a few changes, I can't see the benefit of making explicit something that is implicit in our work. I guess you might suggest it to prevent an automatic regeneration of artefacts without paying attention to their content, but I am still not sure if adding that text would help (after a few weeks the text could also be added without further thought).

@pmanrubia: You're exactly correct when you say, "to prevent an automatic regeneration of artefacts without paying attention to their content". Given that the regression test artefacts are quite painful, I think it's all too easy to regenerate and commit them without thinking too much about the actual change. Having to explain the change forces us to really think about whether it makes sense.

chrisroos commented:
> ... but for comments related to a specific bit of code/commit, it would be handy to add the comment at the diff level?. Something similar to this PR #2502 (diff). It's easy to add a Line note without having to manually include the context.
> Does it make sense for you?

I'm happy to comment on the diff in future, @pmanrubia. Sorry for taking so long to reply to your suggestion/question!
